[TensorExpr] Handle not-specified dtypes and strides.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59347 [TensorExpr] Add NNC lowerings for `aten::mean`, `aten::addmm`, and `aten::adaptive_avg_pool2d`.
* **#59346 [TensorExpr] Handle not-specified dtypes and strides.**
* #59163 [JIT] ShapeProp: add missing ops from mobilenet v3.

Currently JIT has a pass to propagate shapes, but doesn't have a
capability to fill in strides and dtypes. This PR works around that by
assuming default dtype to be Float and strides corresponding to
contiguous layout, unless otherwise specified. Ideally, we won't need
this, and this is done simply as a workaround unless the corresponding
features are implemented on JIT side.

This is required for AOT compilation of mobilenet v3 with NNC.

Differential Revision: [D28853831](https://our.internmc.facebook.com/intern/diff/D28853831)

Commit:
[TensorExpr] Handle not-specified dtypes and strides.

Currently JIT has a pass to propagate shapes, but doesn't have a
capability to fill in strides and dtypes. This PR works around that by
assuming default dtype to be Float and strides corresponding to
contiguous layout, unless otherwise specified. Ideally, we won't need
this, and this is done simply as a workaround unless the corresponding
features are implemented on JIT side.

This is required for AOT compilation of mobilenet v3 with NNC.

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28853831).

navahgar commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -518,7 +524,7 @@ std::vector<ExprHandle> TensorExprKernel::sizesForValue(
   if (v->type()->kind() == TypeKind::TensorType) {
     auto tt = v->type()->cast<TensorType>();
-    if (tt->isComplete()) {
+    if (tt->sizes().isComplete()) {
```
Shouldn't we do `tt->sizes().concrete_sizes()` here instead? It looks like `isComplete()` is checking for more than just if the dims in sizes are not null. See https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/jit_type.h#L445

ZolotukhinM replied:
Correct, but that's intentional in this PR and the justification is in the commit message :)

navahgar replied:
I'm not talking about `tt->isComplete()`. I understand that is not the right thing to do here.

I am referring to `tt->sizes().isComplete()` which is the update you have made. In `VaryingShape`, `concrete_sizes()` only checks if the dims in the shape are not null, which is what we want here. On the other hand, `isComplete()` on `VaryingShape` checks for more, which seems unnecessary here.

navahgar commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -3260,12 +3269,12 @@ void TensorExprKernel::compile() {
     // If the tensor is not dense or overlaps, we have
     // no way of matching the profiled striding
-    if (denseAndNonOverlapping(sizes, strides)) {
-      tensorOutputStrides_.push_back(*tt->strides().concrete_sizes());
```
Just curious: how was this working before? Was this causing a bug when strides are not concrete?

ZolotukhinM replied:
This has never been the case as the only scenario this was used is in the PE, and PE fills in both strides and sizes. In the AOT scenario we currently have a way to populate sizes, but not strides - hence the change.

bertmaher replied:
Why can't we populate strides AOT?

ZolotukhinM replied:
We can, I just felt for some unknown reason that it would be slightly easier this way. Eventually we should have a pass to fill in strides.

Commit:
Update on "[TensorExpr] Handle not-specified dtypes and strides."


Currently JIT has a pass to propagate shapes, but doesn't have a
capability to fill in strides and dtypes. This PR works around that by
assuming default dtype to be Float and strides corresponding to
contiguous layout, unless otherwise specified. Ideally, we won't need
this, and this is done simply as a workaround unless the corresponding
features are implemented on JIT side.

This is required for AOT compilation of mobilenet v3 with NNC.

Differential Revision: [D28853831](https://our.internmc.facebook.com/intern/diff/D28853831)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28853831).

eellison reviewed (commented):
Just throwing this out there, should we consider setting the strides/dtype as part of the AOT pass instead of in the fuser ?

ZolotukhinM commented:
> should we consider setting the strides/dtype as part of the AOT pass instead of in the fuser ?

Do you mean that in the fuser pass we'd allow strides/dtypes to be unspecified? If so, how would TE handle this?

eellison commented:
as part of the AOT pass we would run a pass that would set dtypes to float and strides to contiguous

ZolotukhinM commented:
Ah, I see what you mean. Yes, we probably should do that - for dtypes eventually I hope to use the dtype propagation pass, and for the strides we either could have a separate pass or make it a part of the shape-prop. For now neither of these components exists, and that's why I came up with this temporal workaround.

eellison reviewed (approved):
fine with this, we might reconsider aspects of the AOT pipeline in the future

bertmaher reviewed (approved):
Seems fine, I'm just curious about AOT strides.

eellison commented:
More specifically, we might consider running the graph with PE and using the information generated there, then guarding against it. (I think you've done something similar before)

Commit:
Update on "[TensorExpr] Handle not-specified dtypes and strides."


Currently JIT has a pass to propagate shapes, but doesn't have a
capability to fill in strides and dtypes. This PR works around that by
assuming default dtype to be Float and strides corresponding to
contiguous layout, unless otherwise specified. Ideally, we won't need
this, and this is done simply as a workaround unless the corresponding
features are implemented on JIT side.

This is required for AOT compilation of mobilenet v3 with NNC.

Differential Revision: [D28853831](https://our.internmc.facebook.com/intern/diff/D28853831)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28853831).

Commit:
rebase on "[TensorExpr] Handle not-specified dtypes and strides."


Currently JIT has a pass to propagate shapes, but doesn't have a
capability to fill in strides and dtypes. This PR works around that by
assuming default dtype to be Float and strides corresponding to
contiguous layout, unless otherwise specified. Ideally, we won't need
this, and this is done simply as a workaround unless the corresponding
features are implemented on JIT side.

This is required for AOT compilation of mobilenet v3 with NNC.

Differential Revision: [D28853831](https://our.internmc.facebook.com/intern/diff/D28853831)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28853831).

Commit:
rebase on "[TensorExpr] Handle not-specified dtypes and strides."


Currently JIT has a pass to propagate shapes, but doesn't have a
capability to fill in strides and dtypes. This PR works around that by
assuming default dtype to be Float and strides corresponding to
contiguous layout, unless otherwise specified. Ideally, we won't need
this, and this is done simply as a workaround unless the corresponding
features are implemented on JIT side.

This is required for AOT compilation of mobilenet v3 with NNC.

Differential Revision: [D28853831](https://our.internmc.facebook.com/intern/diff/D28853831)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28853831).

Commit:
rebase on "[TensorExpr] Handle not-specified dtypes and strides."


Currently JIT has a pass to propagate shapes, but doesn't have a
capability to fill in strides and dtypes. This PR works around that by
assuming default dtype to be Float and strides corresponding to
contiguous layout, unless otherwise specified. Ideally, we won't need
this, and this is done simply as a workaround unless the corresponding
features are implemented on JIT side.

This is required for AOT compilation of mobilenet v3 with NNC.

Differential Revision: [D28853831](https://our.internmc.facebook.com/intern/diff/D28853831)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28853831).

Commit:
Update on "[TensorExpr] Handle not-specified dtypes and strides."


Currently JIT has a pass to propagate shapes, but doesn't have a
capability to fill in strides and dtypes. This PR works around that by
assuming default dtype to be Float and strides corresponding to
contiguous layout, unless otherwise specified. Ideally, we won't need
this, and this is done simply as a workaround unless the corresponding
features are implemented on JIT side.

This is required for AOT compilation of mobilenet v3 with NNC.

Differential Revision: [D28853831](https://our.internmc.facebook.com/intern/diff/D28853831)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28853831).

Commit:
rebase on "[TensorExpr] Handle not-specified dtypes and strides."


Currently JIT has a pass to propagate shapes, but doesn't have a
capability to fill in strides and dtypes. This PR works around that by
assuming default dtype to be Float and strides corresponding to
contiguous layout, unless otherwise specified. Ideally, we won't need
this, and this is done simply as a workaround unless the corresponding
features are implemented on JIT side.

This is required for AOT compilation of mobilenet v3 with NNC.

Differential Revision: [D28853831](https://our.internmc.facebook.com/intern/diff/D28853831)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28853831).
